File: csharp/src/Ice/SSL/ConnectorI.cs
// Copyright (c) ZeroC, Inc.

namespace Ice.SSL;

internal sealed class ConnectorI : Ice.Internal.Connector
{
    public Ice.Internal.Transceiver connect() =>
        new TransceiverI(
            _instance,
            _delegate.connect(),
            _host,
            incoming: false,
            serverAuthenticationOptions: null);

    public short type() => _delegate.type();

    public override bool Equals(object obj)
    {
        if (obj is not ConnectorI)
        {
            return false;
        }

        if (this == obj)
        {
            return true;
        }

        var p = (ConnectorI)obj;
        return _delegate.Equals(p._delegate);
    }

    public override int GetHashCode() => _delegate.GetHashCode();

    public override string ToString() => _delegate.ToString();

    // Only for use by EndpointI.
    internal ConnectorI(Instance instance, Ice.Internal.Connector del, string host)
    {
        _instance = instance;
        _delegate = del;
        _host = host;
    }

    private readonly Ice.Internal.Connector _delegate;
    private readonly string _host;
    private readonly Instance _instance;
}

File: java/src/com.zeroc.ice/src/main/java/com/zeroc/Ice/SSL/ConnectorI.java
// Copyright (c) ZeroC, Inc.

package com.zeroc.Ice.SSL;

import com.zeroc.Ice.Connector;
import com.zeroc.Ice.Transceiver;

final class ConnectorI implements Connector {
    @Override
    public Transceiver connect() {
        SSLEngineFactory clientSSLEngineFactory =
                _instance.initializationData().clientSSLEngineFactory;
        if (clientSSLEngineFactory == null) {
            clientSSLEngineFactory =
                    (peerHost, peerPort) ->
                            _instance.engine().createSSLEngine(false, peerHost, peerPort);
        }
        return new TransceiverI(
                _instance, _delegate.connect(), _host, false, clientSSLEngineFactory);
    }

    @Override
    public short type() {
        return _delegate.type();
    }

    @Override
    public String toString() {
        return _delegate.toString();
    }

    @Override
    public int hashCode() {
        return _delegate.hashCode();
    }

    //
    // Only for use by EndpointI.
    //
    ConnectorI(Instance instance, Connector delegate, String host) {
        _instance = instance;
        _delegate = delegate;
        _host = host;
    }

    @Override
    public boolean equals(java.lang.Object obj) {
        if (!(obj instanceof ConnectorI)) {
            return false;
        }

        if (this == obj) {
            return true;
        }

        ConnectorI p = (ConnectorI) obj;
        return p._delegate.equals(_delegate);
    }

    private final Instance _instance;
    private final Connector _delegate;
    private final String _host;
}
